contracts/vortex/CarbonVortex.sol

```diff
@@ -138,109 +345,508 @@ contract CarbonVortex is ICarbonVortex, Upgradeable, ReentrancyGuardUpgradeable,
-            if (balances[i] == 0 || token == _bnt) {
+            uint256 totalFeeAmount = feeAmounts[i];
+            // skip token if no fees have accumulated or token pair is disabled
+            if (totalFeeAmount == 0 || _disabledPairs[token]) {
                 continue;
             }
-
-            // get trade amount for token
-            uint256 tradeAmount;
-            unchecked {
-                // safe because balances[i] >= rewardAmounts[i]
-                tradeAmount = balances[i] - rewardAmounts[i];
+            // skip the final target token
+            if (token == _finalTargetToken) {
+                continue;
+            }
+            // get fee and reward amounts
+            uint256 rewardAmount = rewardAmounts[i];
+            uint256 feeAmount = totalFeeAmount - rewardAmount;
+
+            if (token == _targetToken) {
+                // if _finalTargetToken is not set, directly transfer the fees to the transfer address
+                if (Token.unwrap(_finalTargetToken) == address(0)) {
+                    // safe due to nonReentrant modifier (forwards all gas fees in case of the native token)
+                    _targetToken.unsafeTransfer(_transferAddress, feeAmount);
+                    // increment totalCollected amount
+                    _totalCollected += feeAmount;
+                } else if (
+                    !_tradingEnabled(token) ||
+                    _amountAvailableForTrading(token) < _minTokenSaleAmounts[token] ||
+                    _auctionPriceIsBelowMinimum(token)
+                ) {
+                    // reset trading for target token
+                    _resetTradingTarget(rewardAmount);
+                }
+            } else {
+                if (
+                    !_tradingEnabled(token) ||
+                    _amountAvailableForTrading(token) - feeAmount < _minTokenSaleAmounts[token] ||
+                    _amountAvailableForTrading(token) > _minTokenSaleAmountMultiplier * _minTokenSaleAmounts[token] ||
+                    _auctionPriceIsBelowMinimum(token)
+                ) {
+                    // reset trading for token
+                    _resetTrading(token, rewardAmount);
+                }
             }
+        }
+
+        // allocate rewards to caller
+        _allocateRewards(msg.sender, tokens, rewardAmounts);
+    }
+
+    /**
+     * @notice resets dutch auction for target token -> TKN trades and set the initial price to max possible
+     */
+    function _resetTrading(Token token, uint256 rewardAmount) private {
+        Price memory price = Price({
+            sourceAmount: INITIAL_PRICE_SOURCE_AMOUNT,
+            targetAmount: INITIAL_PRICE_TARGET_AMOUNT
+        });
+        _tradingStartTimes[token] = uint32(block.timestamp);
+        _initialPrice[token] = price;
+        // set min token sale amount
+        _setMinTokenSaleAmount(token, (token.balanceOf(address(this)) - rewardAmount).toUint128() / 2);
```

Comment:
Pretty thin, but dividing by 2 before the cast to `uint128` instead of after, could be the difference between a successful transaction and one which reverts on overflow.

Reply:
imo it's better to revert with overflow if the contract has > uint128 token balance.

Reply:
See [this comment](https://github.com/bancorprotocol/carbon-contracts/pull/140#discussion_r1559754821).